### DISCUSSION TITLE
Swift extension?

### DISCUSSION THREAD

**Discussion by vojto** [tauri-apps/tauri#6223] [Q&A] [answered]
Hey there!

I'd like to build a Tauri app exclusively for macOS. I'm choosing the web stack simply because it's quicker to prototype in.

For this app, I would need to create an extension in Swift. I'll be communicating with some native APIs and such. I already pretty much have the Swift code needed.

I'm hoping for something like [Capacitor plugins](https://capacitorjs.com/docs/plugins/ios).

How could I go about accomplishing this in Tauri?

Thank you!

**Comment by FabianLars** [comment, marked as answer]
So unless you wanna have a localhost server in your swift parts, i guess the only real solution here is https://crates.io/crates/swift-rs.

Edit: forgot to mention the "standard" sidecar approach, using stdio instead of a localhost server. https://tauri.app/v1/guides/building/sidecar

**Reply by vojto** [reply]
I would actually prefer to run a localhost server. The Swift extension would also trigger events for JS to handle.

Does Tauri have a way of starting the server and keeping it alive?

**Reply by FabianLars** [reply]
> Does Tauri have a way of starting the server and keeping it alive?

yeah, look at the sidecar pattern i linked. Your swift app will basically be its own separate app (similar in use to a remote web server), and you can start and stop it as you like.

**Reply by vojto** [reply]
Yeah that makes sense. And I'll have to create a new thread to run it in - [example](https://github.com/tauri-apps/tauri/discussions/2751#discussioncomment-1927078).

Great, I love the approach. I'll have them talk over WebSocket, and I think the total time spent on this will be much less than building native Swift app.

**Comment by MarylnMarconevVe** [comment]
To use native Swift code in a Tauri app, you can create a Tauri plugin. A Tauri plugin is a separate Rust crate that provides additional functionality to your Tauri app. It can be used to interface with native code, like Swift code in your case.

**Reply by vojto** [reply]
That sounds like a different approach, as opposed to running a server. Any pros/cons?

**Reply by semanticist21** [reply]
Seems sidecar pattern more clear and concise
